### DISCUSSION TITLE
Missing DS18x20_o option in ESP32 code

### DISCUSSION THREAD

**Discussion by revk** [arendst/Tasmota#18162] [Ideas]
Is there a reason the DS18X20_o seems to have been dropped in the ESP32 code.

The Shelly Plus 1 finally has an add on, and I would assume the same GPIO0 logic to set the DS18B20 output mode.

Any chance of this being reinstated / added to ESP32 code.

![298150](https://user-images.githubusercontent.com/996983/224477398-3a11ddb2-8616-4e03-a00b-ed0fb3aa9b92.jpg)

**Comment by lalo-uy** [comment]
If you are wiring Vcc in the sensor why not use DS18x20 ?

**Reply by revk** [reply]
DS18x20 GPIO works if the GPIO is a normal direct connection to the DS18B20 data line.

But the add-on has an isolator chip of some sort, as most off these modules run of mains power and float at mains voltage. This means you have to control the GPIO direction with a separate pin, hence the DS18x20_o option on the ESP8266 build.

As I understand it...

**Comment by tjheikki** [comment]
Same here, I have this DS18B20_o listed with a "classic" 1PM + AddOn, working fine, but this choice is missing in Tasmota "12.4.0.2(solo1)single-core" running in a 1PM Plus + the new AddOn. Guess enabling something in source and compiling would do the trick but would obviously good to have it there by default for mortals to use.
1PM Plus + AddOn seems to be currently the only option for an electrically safe remotely controllable thermostat with power monitoring (which one would think every heater would need to have nowadays) so wouldn't consider it niche. Sure there are wireless temp sensors but I'd prefer not to keep changing batteries and have my pipes frozen because of connectivity issues or a Raspberry down.

**Reply by Noschvie** [reply]
@tjheikki which changes have you done to get the DS18B20 running? Thanks

**Reply by tjheikki** [reply]
Haven't got it running with the Plus, only the old one (which aren't available anymore unfortunately).

**Comment by mediarath** [comment]
Here 2 photos of the board if it helps to find a solution:

![1](https://user-images.githubusercontent.com/76083401/229243965-0aed93ae-d639-46bb-88ae-a42848471151.jpg)
![2](https://user-images.githubusercontent.com/76083401/229243975-82bc41a1-182e-4243-970a-aa4c0da31888.jpg)

**Comment by arendst** [comment]
> Is there a reason the DS18X20_o seems to have been dropped in the ESP32 code.

Yes. Timing DS18B20 with the shelly option is a nightmare. To make it work the OneWire library was dissected to add DS18x20_o support. While developing I ran accross many unexplainable errors depending on source code line-up (and resuilting compiler optimizations) !?!?.

I was very pleased to not have to dissect the OneWire lib again for the ESP32.

If anyone wants to add support for this Shelly specific feature feel free to experiment in getting a stable, reproducable, source code friendly driver. Once done I'm glad to add it as a seperate option to the current stable ESP32 DS18x20 driver.

**Comment by tjheikki** [comment]
Thanks for the explanation, Theo. I'll check the code but I'm not experienced in such timing tweaking so better not hold breath. 
BTW noticed the DS18x20_o option does exist in 12.1.1 installed in my two SonOff POWR316D:s, wonder if that was an attempt to use 8266's source and didn't work properly. Guess I could test wiring a Shelly AddOn to one of these.
![image](https://user-images.githubusercontent.com/16526467/230207046-e3bc1a63-fad6-41aa-b261-64ac05efbcb6.png)

**Reply by arendst** [reply]
> wonder if that was an attempt to use 8266's source and didn't work properly

Correct. There were other issues too like ESP interrupts not being handled correctly leading to many false DS18x20 responses.
